readme.md
# Composite Products - Conditional Images for WooCommerce

### What's This?

Mini-extension for [WooCommerce Composite Products](https://woocommerce.com/products/composite-products?aff=46147&cid=5699243) that allows you to create dynamic, multi-layer Composite Product images that respond to option changes. The mini-extension leverages [Scenarios](https://docs.woocommerce.com/document/composite-products/composite-products-advanced-configuration/) to conditionally layer additional images over the main Composite Product image.

### How Does It Work?

The mini-extension introduces a new **Scenario Action** called **Overlay Image**, which requires you to choose an image from your Media library. When the Scenario matching conditions are satified, this image is overlaid on top of the main Composite Product image.

When multiple **Overlay Image** Scenarios exist, each may contribute its own image to the main product image stack. The layering order of these conditional images depends on the relative position of the Scenarios.

This approach makes it possible to utilize either **swapping** or **compositing** techniques to conditionally modify the main product image, depending on your needs:

#### Image Swapping

If you have previously worked with Variable Products and Variation Images, chances are you are already familiar with this concept. Here's the idea behind image swapping:

The main product image is **swapped** with an image that depicts the chosen options as soon as the configuration of the Composite Product is complete.

Instead of replacing the product images, the same effect can be achieved by covering the main product image with a fully opaque (non-transparent) image.

#### Image Compositing

**Compositing** is the combination of multiple layered images into a single image. The final composition creates the illusion that all stacked images are parts of the same scene. For this technique to work, each layered image must include:

* some **transparent** areas that allow lower layers of the stack to be seen;
* some **opaque** areas that introduce new elements in the scene by covering the layers below; and
* possibly, some **semi-trasparent** areas that introduce new elements while partially allowing the layers below to be seen.

#### Swapping vs Compositing

The **Swapping** technique is a viable choice if the number of Component Options is relatively low, as it requires you to **capture one picture** and **create one Scenario** for every possible configuration. Its main advantage over the **Compositing** technique is that it allows you to obtain a natural-looking result, even if you don't have advanced image processing skills.

The **Compositing** technique is often a better choice if the number of possible configurations is high, as it requires you to create one "compositable" image for each Component Option. However, specialized knowledge and tools are required to produce a natural-looking result. For this reason, the Compositing technique is often preferred when 3D product models are available, as 3D modelling software can speed up the production of "compositable" images.

### Examples

#### Image Swapping

The two techniques can be better understood by looking at a practical example. The Build Your Outfit composite product. The product consists of a **T-Shirt**, a pair of **Shoes** and a free **Accessory**.

<img width="1212" alt="build-your-outfit" src="https://user-images.githubusercontent.com/1783726/60343983-33003b80-99be-11e9-9d95-c4de60f34bbb.png">

Consider the following configuration of Component Options:

1. White T-Shirt
2. Brown Shoes
3. Red Tie

Our objective is to swap the main product image once these options have been selected. To achieve this:

1. Create a new Scenario, titled "Swap image: White T-Shirt / Brown Shoes / Red Tie".
2. Locate the **Configuration** section.
3. Tick the **T-Shirt** Component checkbox.
4. Choose **is**.
5. Search for and select the **White T-Shirt**.
6. Tick the **Shoes** Component checkbox.
7. Choose **is**.
8. Search for and select the **Brown Shoes**.
9. Tick the **Accessory** Component checkbox.
10. Choose **is**.
11. Search for and select the **Red Tie**.
12. Locate the **Scenario Actions** section.
13. Check the **Overlay Image** option.
14. Click **Select Image** and choose an image that depicts the White T-Shirt / Brown Shoes / Red Tie configuration.
15. Make sure that the **Activate Options** and **Hide Components** options are disabled.
16. **Update** the product to save your changes.

<img width="788" alt="build-your-outfit-swap-scenario" src="https://user-images.githubusercontent.com/1783726/60343879-f46a8100-99bd-11e9-84bd-329a72c8d010.png">

Now, after choosing the White T-Shirt, Brown Shoes and Red Tie, the main product image will be replaced by the image you selected in Step 14. To achieve the same result when any combination of options is selected, repeat Steps 1-15 for all possible **Build Your Outfit** configurations.

<img width="1209" alt="build-your-outfit-swapped" src="https://user-images.githubusercontent.com/1783726/60343915-0d733200-99be-11e9-9890-4fd7e6fbf292.png">

Note that the image will be updated after **all 3 Component Options have been selected**.

#### Image Compositing

This time, instead of creating an image (and Scenario) for every possible configuration, we will layer multiple images on top of each other to obtain the same end-result. An advantage of this method is that the main product image will be updated **while** the **T-Shirt**, **Shoes** and **Accessory** Components are configured.

Let's start by adding an image for the **White T-Shirt** option:

1. Create a new Scenario, titled "White T-Shirt Image".
2. Locate the **Configuration** section.
3. Tick the **T-Shirt** Component checkbox.
4. Choose **is**.
5. Search for and select the **White T-Shirt**.
6. Leave the **Shoes** and **Accessory** Component boxes unchecked.
7. Locate the **Scenario Actions** section.
8. Check the **Overlay Image** option.
9. Click **Select Image** and choose an image that depicts the **White T-Shirt**.
10. Make sure that the **Activate Options** and **Hide Components** options are disabled.
11. **Update** the product to save your changes.

<img width="788" alt="build-your-outfit-comped-scenario-tee" src="https://user-images.githubusercontent.com/1783726/60344019-47dccf00-99be-11e9-8eae-d09c3d072250.png">

Our objective is to replace the featured image of the product with a new "scene", and add a new element every time a Component Option is selected. To achieve this, the image that we added in Step 9 **completely covers** the original featured image:

<img width="1201" alt="build-your-outfit-comped-tee" src="https://user-images.githubusercontent.com/1783726/60344039-54f9be00-99be-11e9-9404-d22e780b8c20.png">

Next, we will add a "compositable" image for the **Brown Shoes** option:

1. Create a new Scenario, titled "Brown Shoes Image".
2. Locate the **Configuration** section.
3. Tick the **T-Shirt** Component checkbox.
4. Choose **is**.
5. Select the **Any Product or Variation** option.
6. Tick the **Shoes** Component checkbox.
7. Choose **is**.
8. Search for and select the **Brown Shoes**.
9. Leave the **Accessory** Component box unchecked.
10. Locate the **Scenario Actions** section.
11. Check the **Overlay Image** option.
12. Click **Select Image** and choose an image that depicts the **Brown Shoes**.
13. Make sure that the **Activate Options** and **Hide Components** options are disabled.
14. **Update** the product to save your changes.

<img width="788" alt="build-your-outfit-comped-scenario-shoes" src="https://user-images.githubusercontent.com/1783726/60344152-8a061080-99be-11e9-8b1d-a33654481bb0.png">

Note that:

* The Brown Shoes image we selected in Step 12 contains a transparent area that allows the White T-Shirt image to remain visible when the Brown Shoes option is selected.
* The Brown Shoes image should be layered over the chosen T-Shirt image only when a T-Shirt is selected. This is achieved in Steps 3-5.

<img width="1208" alt="build-your-outfit-comped-shoes" src="https://user-images.githubusercontent.com/1783726/60344163-91c5b500-99be-11e9-8640-4ab24c855d97.png">

Finally, let's add a "compositable" image for the **Red Tie** option by repeating steps 1-14 above:

1. Create a new Scenario, titled "Red Tie Image".
2. Locate the **Configuration** section.
3. Tick the **T-Shirt** Component checkbox.
4. Choose **is**.
5. Select the **Any Product or Variation** option.
6. Tick the **Accessory** Component checkbox.
7. Choose **is**.
8. Search for and select the **Brown Shoes**.
9. Leave the **Accessory** Component box unchecked.
10. Locate the **Scenario Actions** section.
11. Check the **Overlay Image** option.
12. Click **Select Image** and choose an image that depicts the **Brown Shoes**.
13. Make sure that the **Activate Options** and **Hide Components** options are disabled.
14. **Update** the product to save your changes.

<img width="787" alt="build-your-outfit-comped-accessory" src="https://user-images.githubusercontent.com/1783726/60344179-9db17700-99be-11e9-95ce-ae111e642064.png">

<img width="1209" alt="build-your-outfit-swapped" src="https://user-images.githubusercontent.com/1783726/60344196-a609b200-99be-11e9-8b21-8f05d3da855b.png">

### Notes

#### Requirements

The mini-extension:

1. Requires Composite Products 5.0+.
2. Only works with themes that declare support for the `wc-product-gallery-slider` feature.

#### Limitations

1. The product image cannot be opened in a lightbox once the mini-extension has stacked additional images over it.
2. The image zooming feature included with WooCommerce is not compatible with the swapping/compositing techniques introduced by the mini-extension. When a Composite Product includes Scenarios with the **Overlay Image** Scenario Action, the image zooming feature of the Composite Product is disabled.




changelog.txt
*** WooCommerce Composite Products - Conditional Images ***

2020.02.07 - version 1.1.1
* Fix - Added support for creating 'overlay_image' scenarios via the REST API.
* Tweak - Declared support for WordPress 5.3 and WooCommerce 3.9.

2019.09.15 - version 1.1.0
* Fix - Compatibility with Composite Products 5.0+. Older versions of Composite Products are no longer supported.

2019.07.26 - version 1.0.2
* Fix - Resolve issue with Composite Products that do not have a product image set.

2019.07.16 - version 1.0.1
 * Fix - Resize stacked images when the viewport width changes.
 * Tweak - Adjust overlay image width based on main 'img' width.

2019.06.27 - version 1.0.0
 * Initial Release.


Gruntfile.js
/* jshint node:true */
module.exports = function( grunt ) {

	'use strict';

	grunt.initConfig({

		// Set folder templates.
		dirs: {
			css:  'assets/css',
			js:   'assets/js'
		},

		// JavaScript linting with JSHint.
		jshint: {
			options: {
				'force': true,
				'boss': true,
				'curly': true,
				'eqeqeq': false,
				'eqnull': true,
				'es3': false,
				'expr': false,
				'immed': true,
				'noarg': true,
				'onevar': true,
				'quotmark': 'single',
				'trailing': true,
				'undef': true,
				'unused': true,
				'sub': false,
				'browser': true,
				'maxerr': 1000,
				globals: {
					'jQuery': false,
					'$': false,
					'Backbone': false,
					'_': false,
					'wp': false,
					'wc_composite_admin_params': false,
					'wc_cp_ci_admin_params': false,
					'$wc_cp_window': false
				},
			},
			all: [
				'Gruntfile.js',
				'<%= dirs.js %>/*.js',
				'!<%= dirs.js %>/*.min.js'
			]
		},

		// Remove jshint headers.
		preprocess : {
			js: {
				src : [
					'<%= dirs.js %>/*js',
					'!<%= dirs.js %>/*.min.js'
				],
				options: {
					inline : true,
				},
			},
		},

		// Minify .js files.
		uglify: {
			options: {
				preserveComments: false
			},
			jsfiles: {
				files: [{
					expand: true,
					cwd: '<%= dirs.js %>',
					src: [
						'*.js',
						'!*.min.js'
					],
					dest: '<%= dirs.js %>',
					ext: '.min.js'
				}]
			}
		},

		// Autoprefixer.
		postcss: {
			options: {
				processors: [
					require( 'autoprefixer' )( {
						browsers: [
							'> 0.1%',
							'ie 8',
							'ie 9'
						]
					} )
				]
			},
			dist: {
				src: [
					'<%= dirs.css %>/**/*.css'
				]
			}
		},

		rtlcss: {
			options: {
				config: {
					swapLeftRightInUrl: false,
					swapLtrRtlInUrl: false,
					autoRename: false,
					preserveDirectives: true
				}
			},
			main: {
				expand: true,
				ext: '-rtl.css',
				src: [
					'<%= dirs.css %>/meta-boxes-product.css',
					'<%= dirs.css %>/single-product.css'
				]
			}
		},

		// Generate POT files.
		makepot: {
			options: {
				type: 'wp-plugin',
				domainPath: 'languages',
				potHeaders: {
					'report-msgid-bugs-to': 'https://woocommerce.com/my-account/create-a-ticket/',
					'language-team': 'LANGUAGE <EMAIL@ADDRESS>'
				}
			},
			go: {
				options: {
					potFilename: 'woocommerce-composite-products-conditional-images.pot',
					exclude: [
						'languages/.*',
						'assets/.*',
						'node-modules/.*',
						'woo-includes/.*'
					]
				}
			}
		},

		// Check textdomain errors.
		checktextdomain: {
			options:{
				text_domain: [ 'woocommerce-composite-products-conditional-images' , 'woocommerce-composite-products' ],
				keywords: [
					'__:1,2d',
					'_e:1,2d',
					'_x:1,2c,3d',
					'esc_html__:1,2d',
					'esc_html_e:1,2d',
					'esc_html_x:1,2c,3d',
					'esc_attr__:1,2d',
					'esc_attr_e:1,2d',
					'esc_attr_x:1,2c,3d',
					'_ex:1,2c,3d',
					'_n:1,2,4d',
					'_nx:1,2,4c,5d',
					'_n_noop:1,2,3d',
					'_nx_noop:1,2,3c,4d'
				]
			},
			files: {
				src:  [
					'**/*.php', // Include all files
					'!apigen/**', // Exclude apigen/
					'!deploy/**', // Exclude deploy/
					'!node_modules/**' // Exclude node_modules/
				],
				expand: true
			}
		}
	});

	// Load NPM tasks to be used here
	grunt.loadNpmTasks( 'grunt-preprocess' );
	grunt.loadNpmTasks( 'grunt-contrib-jshint' );
	grunt.loadNpmTasks( 'grunt-contrib-uglify' );
	grunt.loadNpmTasks( 'grunt-postcss' );
	grunt.loadNpmTasks( 'grunt-rtlcss' );
	grunt.loadNpmTasks( 'grunt-wp-i18n' );
	grunt.loadNpmTasks( 'grunt-checktextdomain' );

	// Register tasks.
	grunt.registerTask( 'dev', [
		'checktextdomain',
		'uglify',
		'preprocess',
		'postcss',
		'rtlcss'
	] );

	grunt.registerTask( 'default', [
		'dev',
		'makepot'
	] );
};


assets/js/single-product.js
;( function ( $, window, document, Backbone, undefined ) {

	$( '.composite_data' ).on( 'wc-composite-initializing', function( event, composite ) {

		function Overlay_Image_App() {

			/**
			 * Track changes to 'Overlay Image' scenarios.
			 */
			var Overlay_Image_Model = function( opts ) {

				var Model = Backbone.Model.extend( {

					initialize: function( options ) {

						composite.actions.add_action( 'active_scenarios_updated', this.component_selection_changed_handler, 20, this );

						var params = {
							active_scenarios: [],
						};

						this.set( params );
					},

					component_selection_changed_handler: function( step ) {

						var active_scenarios = composite.scenarios.get_active_scenarios_by_type( 'overlay_image' )

						this.set( { active_scenarios: active_scenarios } );
					}

				} );

				var obj = new Model( opts );
				return obj;
			};

			/**
			 * Render overlays.
			 */
			var Overlay_Image_View = function( opts ) {

				var View = Backbone.View.extend( {

					$main_image_container: false,
					$main_image: false,

					initialize: function( options ) {

						this.$main_image_container = this.$el.find( '.woocommerce-product-gallery__image' ).first();

						if ( this.$main_image_container.length > 0 ) {
							this.$main_image = this.$main_image_container.find( 'a img' ).first();
						} else {
							this.$main_image_container = this.$el.find( '.woocommerce-product-gallery__image--placeholder' ).first();
							if ( this.$main_image_container.length > 0 ) {
								this.$main_image = this.$main_image_container.find( 'img' ).first();
							}
						}

						if ( ! this.$main_image ) {
							return;
						}

						this.listenTo( this.model, 'change:active_scenarios', this.render );

						var view = this;

						/**
						 * Recalculate overlay widths on resize.
						 */
						$wc_cp_window.resize( function() {

							if ( ! composite.is_initialized ) {
								return false;
							}

							view.handle_resize();

						} );

						if ( this.$main_image.get(0).complete ) {
							setTimeout( function() {
								view.handle_resize();
							}, 100 );
						} else {
							this.$main_image.one( 'load', function() {
								setTimeout( function() {
									view.handle_resize();
								}, 100 );
							} );
						}
					},

					render: function() {

						var active_scenarios = this.model.get( 'active_scenarios' ),
							image_width      = this.$main_image.width(),
							overlay_css      = { width: image_width, position: 'absolute', 'z-index': 1 };

						// Remove overlays.
						this.$main_image_container.find( '.wc-cp-overlay-image' ).remove();

						for ( var index = active_scenarios.length - 1; index >= 0; index-- ) {

							var scenario_id = active_scenarios[ index ],
								image_html  = composite.scenarios.get_scenario_data().scenario_settings.overlay_image[ scenario_id ],
								$image_html = $( image_html ).css( overlay_css );

							if ( image_html ) {
								this.$main_image_container.prepend( $image_html );
							}
						}
					},

					handle_resize: function() {

						var image_width = this.$main_image.width(),
							overlay_css = { width: image_width };

						this.$main_image_container.find( '.wc-cp-overlay-image' ).css( overlay_css );
					}

				} );

				var obj = new View( opts );
				return obj;
			};

			/**
			 * Initialize app.
			 */
			this.initialize = function() {

				var $images_wrapper = $( '.woocommerce-product-gallery__wrapper' ).first();

				if ( $images_wrapper.length > 0 ) {
					this.model = new Overlay_Image_Model();
					this.view = new Overlay_Image_View( { model: this.model, el: $images_wrapper } );
				}

			};
		}

		var app = new Overlay_Image_App();

		app.initialize();

	} );

} ) ( jQuery, window, document, Backbone );


assets/js/meta-boxes-product.js
;( function( $ ) {

	$( function( $ ) {

		var $scenarios_panel             = $( '#bto_scenario_data' ),
			conditional_image_frame_data = {
				image_frame: false,
				$button:     false
			};

		/*
		 * Handle events in Scenarios panel.
		 */

		// Update scenario DOM elements, menu order and toolbar state.
		$scenarios_panel

			// Enable/Disable Scenario Action.
			.on( 'change', 'input.scenario_action_overlay_image', function() {

				var $this      = $( this ),
					$container = $this.closest( '.scenario_action_overlay_image_group' ),
					$content   = $container.find( '.action_conditional_images' );

				if ( $this.is( ':checked' ) ) {
					$content.show();
				} else {
					$content.hide();
				}

			} )

			// Set Image.
			.on( 'click', '.upload_conditional_image_button', function( e ) {

				conditional_image_frame_data.$button = $( this );

				e.preventDefault();

				// If the media frame already exists, reopen it.
				if ( conditional_image_frame_data.image_frame ) {

					conditional_image_frame_data.image_frame.open();

				} else {

					// Create the media frame.
					conditional_image_frame_data.image_frame = wp.media( {

						// Set the title of the modal.
						title: wc_cp_ci_admin_params.i18n_choose_component_image,
						button: {
							text: wc_cp_ci_admin_params.i18n_set_component_image
						},
						states: [
							new wp.media.controller.Library( {
								title: wc_cp_ci_admin_params.i18n_choose_component_image,
								filterable: 'all'
							} )
						]
					} );

					// When an image is selected, run a callback.
					conditional_image_frame_data.image_frame.on( 'select', function () {

						var attachment = conditional_image_frame_data.image_frame.state().get( 'selection' ).first().toJSON(),
							url        = attachment.sizes && attachment.sizes.thumbnail ? attachment.sizes.thumbnail.url : attachment.url;

						conditional_image_frame_data.$button.addClass( 'has_image' );
						conditional_image_frame_data.$button.closest( '.action_conditional_images' ).find( '.remove_conditional_image_button' ).addClass( 'has_image' );
						conditional_image_frame_data.$button.find( 'input' ).val( attachment.id ).change();
						conditional_image_frame_data.$button.find( 'img' ).eq( 0 ).attr( 'src', url );
					} );

					// Finally, open the modal.
					conditional_image_frame_data.image_frame.open();
				}

			} )

			// Remove Image.
			.on( 'click', '.remove_conditional_image_button', function( e ) {

				var $button         = $( this ),
					$option_wrapper = $button.closest( '.action_conditional_images' ),
					$upload_button  = $option_wrapper.find( '.upload_conditional_image_button' );

				e.preventDefault();

				$upload_button.removeClass( 'has_image' );
				$button.removeClass( 'has_image' );
				$option_wrapper.find( 'input' ).val( '' ).change();
				$upload_button.find( 'img' ).eq( 0 ).attr( 'src', wc_composite_admin_params.wc_placeholder_img_src );

			} );

	} );

} ) ( jQuery );


composite-products-conditional-images-for-woocommerce.php
<?php
/**
* Plugin Name: Composite Products - Conditional Images for WooCommerce
* Plugin URI: https://docs.woocommerce.com/document/composite-products/composite-products-extensions/#cp-ci
* Description: Free mini-extension for WooCommerce Composite Products that allows you to create dynamic, multi-layer Composite Product images that respond to option changes.
* Version: 1.2.0
* Author: SomewhereWarm
* Author URI: https://somewherewarm.gr/
*
* Text Domain: woocommerce-composite-products-conditional-images
* Domain Path: /languages/
*
* Requires at least: 4.4
* Tested up to: 5.3
*
* WC requires at least: 3.1
* WC tested up to: 3.9
*
* Copyright: © 2017-2020 SomewhereWarm SMPC.
* License: GNU General Public License v3.0
* License URI: http://www.gnu.org/licenses/gpl-3.0.html
*/

// Exit if accessed directly.
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Main plugin class.
 *
 * @class    WC_CP_Conditional_Images
 * @version  1.2.0
 */
class WC_CP_Conditional_Images {

	/**
	 * Plugin version.
	 *
	 * @var string
	 */
	public static $version = '1.2.0';

	/**
	 * Min required CP version.
	 *
	 * @var string
	 */
	public static $req_cp_version = '5.0';

	/**
	 * Plugin URL.
	 *
	 * @return string
	 */
	public static function plugin_url() {
		return plugins_url( basename( plugin_dir_path(__FILE__) ), basename( __FILE__ ) );
	}

	/**
	 * Plugin path.
	 *
	 * @return string
	 */
	public static function plugin_path() {
		return untrailingslashit( plugin_dir_path( __FILE__ ) );
	}

	/**
	 * Fire in the hole!
	 */
	public static function init() {
		add_action( 'plugins_loaded', array( __CLASS__, 'load_plugin' ) );
	}

	/**
	 * Hooks.
	 */
	public static function load_plugin() {

		if ( ! function_exists( 'WC_CP' ) || version_compare( WC_CP()->version, self::$req_cp_version ) < 0 ) {
			add_action( 'admin_notices', array( __CLASS__, 'cp_version_check_notice' ) );
			return false;
		}

		// Localization.
		add_action( 'init', array( __CLASS__, 'localize_plugin' ) );

		// Front-end script (where the magic happens).
		add_filter( 'woocommerce_composite_script_dependencies', array( __CLASS__, 'frontend_script' ) );

		// If overlays exist, dequeue zoom script.
		add_action( 'woocommerce_composite_add_to_cart', array( __CLASS__, 'dequeue_zoom_script' ), 10 );

		// Admin script.
		add_action( 'admin_enqueue_scripts', array( __CLASS__, 'admin_scripts' ) );

		// Add 'Overlay Image' action in Scenarios.
		add_action( 'woocommerce_composite_scenario_admin_actions_html', array( __CLASS__, 'scenario_admin_actions_html' ), 20, 4 );

		// Save 'Overlay Image' action settings.
		add_filter( 'woocommerce_composite_process_scenario_data', array( __CLASS__, 'process_scenario_data' ), 10, 5 );

		// Add qty data in scenarios.
		add_filter( 'woocommerce_composite_current_scenario_data', array( __CLASS__, 'scenario_data' ), 10, 4 );

		// Allow 'overlay_image' scenario actions to be created via the REST API.
		add_filter( 'woocommerce_rest_api_extended_composite_scenarios_field_args', array( __CLASS__, 'add_rest_api_scenario_action' ) );
	}

	/**
	 * CP version check notice.
	 */
	public static function cp_version_check_notice() {
	    echo '<div class="error"><p>' . sprintf( __( '<strong>Composite Products &ndash; Conditional Images</strong> requires WooCommerce Composite Products <strong>%s</strong> or higher.', 'woocommerce-composite-products-conditional-images' ), self::$req_cp_version ) . '</p></div>';
	}

	/**
	 * Load textdomain.
	 *
	 * @return void
	 */
	public static function localize_plugin() {
		load_plugin_textdomain( 'woocommerce-composite-products-conditional-images', false, dirname( plugin_basename( __FILE__ ) ) . '/languages/' );
	}

	/**
	 * Front-end script.
	 *
	 * @param array $dependencies
	 */
	public static function frontend_script( $dependencies ) {

		if ( ! current_theme_supports( 'wc-product-gallery-slider' ) ) {
			return false;
		}

		$suffix = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '' : '.min';

		wp_register_script( 'wc-add-to-cart-composite-ci', self::plugin_url() . '/assets/js/single-product' . $suffix . '.js', array( 'jquery', 'underscore', 'backbone' ), self::$version );
		$dependencies[] = 'wc-add-to-cart-composite-ci';

		return $dependencies;
	}

	/**
	 * Dequeue zoom script.
	 */
	public static function dequeue_zoom_script( $dependencies ) {

		global $product;

		if ( ! current_theme_supports( 'wc-product-gallery-slider' ) ) {
			return false;
		}

		$has_overlay_image_scenarios = false;
		$scenario_metadata           = $product->get_scenario_data();

		if ( ! empty( $scenario_metadata ) ) {
			foreach ( $scenario_metadata as $scenario_id => $metadata ) {
				if ( isset( $metadata[ 'scenario_actions' ][ 'overlay_image' ][ 'is_active' ] ) && 'yes' === $metadata[ 'scenario_actions' ][ 'overlay_image' ][ 'is_active' ] ) {
					$has_overlay_image_scenarios = true;
					break;
				}
			}
		}

		if ( $has_overlay_image_scenarios ) {
			wp_dequeue_script( 'zoom' );
		}
	}

	/**
	 * Admin scripts and styles.
	 *
	 * @return void
	 */
	public static function admin_scripts() {

		$suffix = defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ? '' : '.min';

		$script_dependency = 'wc-composite-admin-product-panel';
		wp_register_script( 'wc-cp-ci-admin-product-metaboxes', self::plugin_url() . '/assets/js/meta-boxes-product' . $suffix . '.js', array( 'jquery', 'jquery-ui-datepicker', 'wp-util', 'wc-admin-product-meta-boxes', $script_dependency ), self::$version );

		$style_dependency = 'wc-composite-writepanel-css';
		wp_register_style( 'wc-cp-ci-admin-product-metaboxes-css', self::plugin_url() . '/assets/css/meta-boxes-product.css', array( 'woocommerce_admin_styles', $style_dependency ), self::$version );
		wp_style_add_data( 'wc-cp-ci-admin-product-metaboxes-css', 'rtl', 'replace' );

		// Get admin screen id.
		$screen    = get_current_screen();
		$screen_id = $screen ? $screen->id : '';

		if ( 'product' === $screen_id ) {
			wp_enqueue_script( 'wc-cp-ci-admin-product-metaboxes' );

			$params = array(
				'i18n_choose_component_image' => __( 'Choose an Image', 'woocommerce-composite-products-conditional-images' ),
				'i18n_set_component_image'    => __( 'Save Image', 'woocommerce-composite-products-conditional-images' )
			);

			wp_localize_script( 'wc-cp-ci-admin-product-metaboxes', 'wc_cp_ci_admin_params', $params );
		}

		if ( in_array( $screen_id, array( 'edit-product', 'product' ) ) ) {
			wp_enqueue_style( 'wc-cp-ci-admin-product-metaboxes-css' );
		}
	}

	/**
	 * Save qty override settings in scenarios
	 *
	 * @param  array  $scenario_meta
	 * @param  array  $scenario_post_data
	 * @param  string $scenario_id
	 * @param  array  $composite_meta
	 * @param  string $composite_id
	 * @return array
	 */
	public static function process_scenario_data( $scenario_meta, $scenario_post_data, $scenario_id, $composite_meta, $composite_id ) {

		$is_active = ! empty( $scenario_post_data[ 'scenario_actions' ][ 'overlay_image' ][ 'is_active' ] );
		$image_id  = ! empty( $scenario_post_data[ 'scenario_actions' ][ 'overlay_image' ][ 'image_id' ] ) ? $scenario_post_data[ 'scenario_actions' ][ 'overlay_image' ][ 'image_id' ] : '';

		// Save defaults.
		$scenario_meta[ 'scenario_actions' ][ 'overlay_image' ][ 'is_active' ] = 'no';
		$scenario_meta[ 'scenario_actions' ][ 'overlay_image' ][ 'image_id' ] = '';

		// Save image src.
		if ( $image_id ) {
			$scenario_meta[ 'scenario_actions' ][ 'overlay_image' ][ 'image_id' ] = $image_id;
		}

		// Save active state.
		if ( $is_active && $image_id ) {
			$scenario_meta[ 'scenario_actions' ][ 'overlay_image' ][ 'is_active' ] = 'yes';
		}

		return $scenario_meta;
	}

	/**
	 * Add qty data in scenarios
	 *
	 * @param  array  $scenario_data
	 * @param  array  $component_options
	 * @param  string $composite
	 * @return array
	 */
	public static function scenario_data( $scenario_data, $component_options, $composite ) {

		if ( ! current_theme_supports( 'wc-product-gallery-slider' ) ) {
			return $scenario_data;
		}

		$scenario_metadata = $composite->get_scenario_data();

		if ( ! empty( $scenario_data[ 'scenarios' ] ) ) {

			foreach ( $scenario_data[ 'scenarios' ] as $scenario_id ) {

				if ( isset( $scenario_metadata[ $scenario_id ][ 'scenario_actions' ][ 'overlay_image' ][ 'is_active' ] ) && 'yes' === $scenario_metadata[ $scenario_id ][ 'scenario_actions' ][ 'overlay_image' ][ 'is_active' ] ) {

					$image_id = ! empty( $scenario_metadata[ $scenario_id ][ 'scenario_actions' ][ 'overlay_image' ][ 'image_id' ] ) ? $scenario_metadata[ $scenario_id ][ 'scenario_actions' ][ 'overlay_image' ][ 'image_id' ] : '';

					if ( ! $image_id ) {
						continue;
					}

					$image_size = apply_filters( 'woocommerce_gallery_image_size', 'woocommerce_single' );
					$image      = wp_get_attachment_image( $image_id, $image_size, false, apply_filters( 'woocommerce_gallery_image_html_attachment_image_params', array(
						'class'    => 'wp-post-image wc-cp-overlay-image'
					), $image_id, $image_size, true ) );

					$scenario_data[ 'scenario_settings' ][ 'overlay_image' ][ $scenario_id ] = $image;
				}
			}
		}

		return $scenario_data;
	}

	/**
	 * Add 'Overlay Image' action in Scenarios.
	 *
	 * @return void
	 */
	public static function scenario_admin_actions_html( $id, $scenario_data, $composite_data, $product_id ) {

		$overlay_image = isset( $scenario_data[ 'scenario_actions' ][ 'overlay_image' ][ 'is_active' ] ) && 'yes' === $scenario_data[ 'scenario_actions' ][ 'overlay_image' ][ 'is_active' ] ? 'yes' : 'no';
		$image_id      = ! empty( $scenario_data[ 'scenario_actions' ][ 'overlay_image' ][ 'image_id' ] ) ? $scenario_data[ 'scenario_actions' ][ 'overlay_image' ][ 'image_id' ] : '';
		$image         = $image_id ? wp_get_attachment_thumb_url( $image_id ) : '';

		?>
		<div class="scenario_action_overlay_image_group" >
			<div class="form-field toggle_overlay_image">
				<label for="scenario_action_overlay_image_<?php echo $id; ?>">
					<?php echo __( 'Overlay Image', 'woocommerce-composite-products-conditional-images' ); ?>
				</label>
				<input id="scenario_action_overlay_image_<?php echo $id; ?>" type="checkbox" class="checkbox scenario_action_overlay_image" <?php echo ( $overlay_image === 'yes' ? ' checked="checked"' : '' ); ?> name="bto_scenario_data[<?php echo $id; ?>][scenario_actions][overlay_image][is_active]" <?php echo ( $overlay_image === 'yes' ? ' value="1"' : '' ); ?> /><?php
					echo wc_help_tip( __( 'Enable this option to conditionally overlay an image over the main Composite Product image. When using this feature, product image zooming will be disabled.', 'woocommerce-composite-products-conditional-images' ) );
				?>
			</div>
			<div class="form-field action_conditional_images" <?php echo ( $overlay_image === 'no' ? ' style="display:none;"' : '' ); ?> >
				<a href="#" class="upload_conditional_image_button <?php echo $image_id ? 'has_image': ''; ?>">
					<span class="prompt"><?php echo __( 'Select image', 'woocommerce-composite-products' ); ?></span>
					<img src="<?php if ( ! empty( $image ) ) echo esc_attr( $image ); else echo esc_attr( wc_placeholder_img_src() ); ?>" />
					<input type="hidden" name="bto_scenario_data[<?php echo $id; ?>][scenario_actions][overlay_image][image_id]" class="image" value="<?php echo $image_id; ?>" />
				</a>
				<?php echo wc_help_tip( __( 'Choose an image to overlay over the main Composite Product image.', 'woocommerce-composite-products-conditional-images' ) ); ?>
				<a href="#" class="remove_conditional_image_button <?php echo $image_id ? 'has_image': ''; ?>"><?php echo __( 'Remove image', 'woocommerce-composite-products' ); ?></a>
			</div>
		</div>
		<?php
	}

	/**
	 * Add support for creating 'overlay_image' scenario actions via the REST API.
	 *
	 * @since  1.1.1
	 * @param  array
	 */
	public static function add_rest_api_scenario_action( $args ) {
		$args[ 'schema' ][ 'items' ][ 'properties' ][ 'actions' ][ 'items' ][ 'properties' ][ 'action_id' ][ 'enum' ][] = 'overlay_image';
		return $args;
	}
}

WC_CP_Conditional_Images::init();
